### PR TITLE
fixed json decoding

### PR DESCRIPTION
**Pull request by interestIngc**
## Description

Replaced ```json.NewDecoder(r.Body).Decode(&req)``` with ```json.Unmarshal(body, to)```, because ```json.NewDecoder``` does not return any error when the request json is malformed (e.g. input contains some trailing data after the json).

### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
- [x] http://localhost:5001/ home page loads
- [x] "LoginAsUser" button takes you to the user login page
- [x] Log in using "tester" and "12341234" should succeed
- [x] Updating the display name should work
- [x] Log out and Register a new user and try to log in with that new user
- [x] "LoginAsClient" button takes you to the client login page
- [x] Log in using "layer8" and "12341234" should work
- [x] You will be able to see your UID and Secret in profile
- [x] Log out and Register a new client and log in using that should work

### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
- [x] Log in with 'tester' / '1234'
- [x] Log in anonymously with Layer8
- [x] Clicking "Get Next Poem" loads different poem correctly x 3
- [x] Clicking "Logout" takes you to the login screen
- [x] Clicking "Register" takes you to the registration page
- [x] Registering with a username, password, and profile image is successful
- [x] Logging in with the new username / password succeeds
- [x] Logging in with Layer8 opens the pop-up
- [x] Logging in with the newly registered credentials from Section 1 succeeds
- [x] User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
- [x] Clicking "Get Next Poem" loads different poem correctly x 3
- [x] Clicking "Logout" takes you to the login page
- [x] Log in with 'tester' / '1234', and then with the credentials from Section 1 succeeds
- [x] This time, DO NOT share "display name" or "country"
- [x] Clicking "Get Next Poem" loads different poem correctly x 3

### SECTION 3: IMSHARER (http://localhost:5174/)
- [x] Main page loads
- [x] Upload of image works
- [x] Reload leads to instant reload (demonstrating proper caching)
- [x] Clicking the newly loaded image shows it in a lightbox

### CODE REVIEW

File: server/resource_server/controller/controller.go [hunk 1]
```diff
@@ -62,13 +62,14 @@ func ServeFileHandler(w http.ResponseWriter, r *http.Request, filePath string) {
 
 func LoginClientHandler(w http.ResponseWriter, r *http.Request) {
 	newService := r.Context().Value("service").(interfaces.IService)
-	var req dto.LoginClientDTO
-	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
-		utils.HandleError(w, http.StatusBadRequest, "Failed to get client profile", err)
+
+	var request dto.LoginClientDTO
+	err := decodeJson(w, r.Body, &request)
```
Comment (ternakkode) [inline server/resource_server/controller/controller.go:67]: we can simplify this using generic

File: server/resource_server/controller/controller.go [hunk 2]
```diff
@@ -427,3 +407,25 @@ func CheckBackendURI(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func decodeJson(w http.ResponseWriter, byteBuffer io.ReadCloser, to any) error {
```
Comment (ternakkode) [inline server/resource_server/controller/controller.go:411]: put this function into utils

File: server/resource_server/controller/controller.go (hunk 1 above)
Comment (interestIngc) [reply server/resource_server/controller/controller.go]: Makes sense! Made this function generic

**Review by ternakkode — APPROVED** [review]

**Review by huzaifamk — APPROVED** [review]
Looks good to me, nice work @interestIngc , I have seen the use of this: `[T any]` for the first time, great work!

**Pull request merged**